### DISCUSSION TITLE
AWS-CDK example and some learnings

### DISCUSSION THREAD

**Discussion by pfried** [vercel/turborepo#494] [Ideas]
Hello there,

I do not know if this is of interest to anyone, but I did some `aws-cdk` example with a rather complex pipeline dependency tree. I would propably not do this for a production app this way (I do use aws-cdk pipelines for that purpose) but it was some fiddling around with turborepo for me.

The code is located here: https://github.com/Nantis-GmbH/turborepo/tree/aws-cdk-example/examples/aws-cdk

The example consists of 3 parts:

1. backend (AWS) with a simple API based on a lambda 
2. frontend-hosting (AWS S3)
3. frontend based on vite.js react

My focus was not so much the caching part and the `deploy` tasks are not cacheable anyways (because you will need to make the cdk call in order to make sure the same is deployed to aws as what you have in your `cdk.out` synthezied folder) 

So the tricky part is that the frontend will need to know the API endpoint at build time. So the following pipeline will be needed:

1. build all packages
2. synth and deploy the backend
3. inject the outputs (api endpoint url) to the frontend build task
4. build the frontend
5. synth and deploy the frontend-hosting including the built frontend

_The inject outputs part in 3. is a nasty `cp aws-config.js` script which tightly couples everything - How to do better? imho this should work through node package resolution, but I do not know how to make this work or you will need to include `cdk` as a dependency also in the frontend and then call aws for the `frontend-hosting` stack outputs and use that._

The main issue I am having is that the depency graph is rather cluttered since some of the dependencies are 1 to 1. Is there any way to include a single package dependency in the pipelines? (I tried scope but was not able to make it work correctly). See graph picture below

![graph](https://user-images.githubusercontent.com/1744875/148042943-598c9b92-97e5-4f24-bc9b-6a7ea0a03dde.jpg)

**Comment by jaredpalmer** [comment]
cc AWS crewwwwww! @medelman17 @adamelmore @ericclemmons

**Comment by adamdotdevin** [comment]
I have a ton of thoughts, but I think it'd be easiest for me to explain my approach once I've buttoned up the repo I'm putting together.

Fwiw, my frontend is Next.js deployed to Vercel. So, my GH Action for deploys looks like:
- deploy meta CDK stack that has some resources that should only be deployed into any given account exactly once (GitHub OIDC Provider, some secrets, etc.)
- deploy api CDK stack
- create `.env.local` in the frontend repo through a script that pulls from the CDK outputs of the previous api deploy
- deploy frontend Next.js app using the Vercel CLI

**Reply by adamdotdevin** [reply]
I should have added that the api stack can be deployed into a given AWS account multiple times; this is to facilitate preview builds for PRs into main.

**Reply by pfried** [reply]
From my experience the part of 
> create .env.local in the frontend repo through a script that pulls from the CDK outputs of the previous api deploy

is kind of not really solved though it must be a standard use case ([Link to question on SO](https://stackoverflow.com/questions/65127311/how-to-pass-configuration-to-a-webapp-used-in-a-aws-cdk-stack))

Of course it always depends on what you are using, but programatically pulling data into a build process in a more integrated way would be much better. Are there any "hooks" or something like that you are aware of?

**Comment by jaredpalmer** [comment]
My hunch is that we need to add some kind of PreExec and PostExec commands to `pipeline` entries so that you can alter the hash ahead of time and easily suspend execution until health check passes or something.

**Reply by adamdotdevin** [reply]
For my case, I don't really need a mechanism in turbo to pause as I'm switching over to Vercel for the web deploy. Most of the orchestration is handled through GH actions. My case might be atypical, though, and I'm sure someone would get value out of the proposed features.

**Reply by pfried** [reply]
`aws-cdk pipelines` have those but it is for halting a deployment pipeline for organizational (approval, e.g. trunk based development, deploy to main) or test reasons (integration tests) or manual pause.

**Comment by pfried** [comment]
I guess there are two main directions people are coming from: Either Vercel (nextjs app, Github Actions, few AWS resources) or AWS (Codebuild / cdk-pipelines, many services, maybe Vercel or AWS Amplify for Webapp).

My example is as I said a fiddle. So I will draw a bigger picture here:

Our product are IoT Applications where a lot of devices send data to AWS. The main components (not a monorepo currently, we have a lot of private npm packages which I am trying to change) are:

1. Backend running on AWS (hundreds of AWS resources, a lot of Lambdas)
2. User Frontend (React app, statically hosted)
3. Marketing Site (Next.js app, possible on vercel or aws)
4. Ideally a separate storybook app since this just clutters the other apps otherwise

The apps need to share packages therefore they include their dependencies (currently all private npm packages)

If it comes to deployment / CI / CD we use [aws-cdk pipelines](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.pipelines-readme.html)

The deployment consists of multiple steps (build, test, synth, deploy and also stages (dev, prod). So the way turborepo fits in is more the local development of the webapps and building packages. Within a deployment turborepo tasks will be run by an overlying pipeline or github actions anyway as they need to happen event based.

To sum it up:

- Turborepo is a tool below an orchestration level. If you do not use CI/CD you can it on that level tough (see my example)
- Linking a backend to a frontend kind of sucks
- The nature of a monorepo adds a ton of issues which is more the fault of a specific package manager or library (e.g. hoisting with npm causes issues in our CRA app)
- Caching of synthesized resources is possible
- Caching of deployed cloud resources is not possible when systems do not share state (maybe solvable on a vercel level, but for e.g. aws-cdk it needs to check if the cloudformation templates and others match the deployed state).
- It would help if the graphing tool only included scripts which are present in the `package.json`. Otherwise there will be quite some "empty" depencies in there

**Reply by JohnGrisham** [reply]
I wasn't aware of pipelines and I'm fairly new to the CDK in general but I like the overall approach and I think it fits with what I'm building. As for the second item in your summary about linking the frontend with backend code I've found this to be true as well in Nx.

When deleting/re-deploying your stack there needs to be some kind of resyncing process for the various resources that get used by your frontend deployment without having to completely redeploy the frontend code. I've considered maybe making the glue be some configuration file in s3 that holds the arns/ids for whatever resources you need to reference on the frontend. Although I don't really like this approach either because it poses security risks especially when it comes to things like user pool ids.

Regardless, I've continued to use Nx for my AWS monorepo mostly because of the codegen support. I'm now aware of plop as an alternative for Turborepo but would still prefer something built-in.

**Reply by pfried** [reply]
When we redeploy the frontend deploy phase will always be hit as it is part of the pipeline, so that is solved. If you see them as independent resources you need to make sure it gets resynced (which means it should just run again). The frontend deployment is actually very "cheap" as it runs quickly (at least that is true for a typical webapp).

The frontend deploy gets its configuration (say API endpoint) via an [SDK call from Cloudformation](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-cloudformation/classes/describestackscommand.html) and then uses it (how depends on your tooling) to build the app. So the deploy frontend action can be called inside or outside of a CI pipeline and produces the same result.

**Comment by ashwinchandran13** [comment]
I'm not sure if this fits with this writeup but I'm setting up a React+Vite app in turborepo with CI/CD setup in AWS Codepipeline i.e. CodeCommit, CodeBuild, CodeDeploy. While coming across the CodeBuild buildspec.yml, I saw a directive of caching. So I was just thinking if we could make use of turborepo's capability of caching into CodeBuild i.e. we can store .cache/turbo in an S3 bucket and allow codebuild to fetch this cache use it.
![Uploading image.png…]()
